docs/StandardLibraryProgrammersManual.md

```diff
@@ -44,6 +44,20 @@ TODO: Should this subsume or link to [AccessControlInStdlib.rst](https://github.
 1. Frequently Encountered Issues
 
 
+## Coding style
+
+- The stdlib currently has a hard line length limit of 80 characters. 
+
+   To break long lines, please closely follow the indentation conventions you see in the existing codebase. (FIXME: Describe.)
+  
+- All entities that aren't public API must include at least one underscored component in their fully qualified names. This includes symbols that are technically declared public but that aren't considered part of the public stdlib API, as well as `@usableFromInline internal`, plain `internal` and `[file]private` types and members. 
+
+    The underscored component need not be the last. For example, `Swift.Dictionary._worble()` is a good name for an internal helper method, but so is `Swift._NativeDictionary.worble()` -- the `_NativeDictionary` type already has the underscore.
+    
+    Initializers don't have a dedicated name we can put the underscore on; instead, we add the underscore on the first argument label: `init(_value: Int)`. If the initializer doesn't have any parameter, then we add a dummy parameter of type Void with an underscored label: for example, `UnsafeBufferPointer.init(_empty: ())`.
+
+    This rule ensures we don't accidentally clutter the public namespace with `@usableFromInline` things (which could prevent us from choosing the best names for new public API), and it also makes it easy to see at a glance if a piece of stdlib code uses any non-public entities.
+
```

Comment:
Why? In general, or only if it's part of the ABI?

Reply:
When the underscoring rule was introduced, everything in the stdlib was effectively inlinable. 

This is no longer the case, but I see no reason to cease applying the rule. I find the underscores are a useful signal when I read the code.

On a practical level, it would be all too easy to accidentally slap on `@usableFromInlne` on an un-underscored internal method without renaming it.

Reply:
Do members of internal types also need an underscore?

One the one hand, an internal type `_Foo` may become public at one point, but we want its ABI imprint to remain. We'd change `_Foo` to public and say `public typealias Foo = _Foo`. At that point, it has many violations of the underscore rule. On the other hand, I don't think this is the sort of scenario to design an entire convention around.

Reply:
Internal types need to have an underscore in the type name. The members need not have an underscore themselves (and often cannot, because of protocol conformances).

The proper way to make a previously `@usableFromInline internal struct _Foo` public would be to keep the existing type and to wrap it in a new public struct. (Unless the internal API happens to be accepted verbatim as public. But that is exceedingly unlikely.)

To make a previously internal/private (non-usableFromInline) type public, we remove the old type and replace it wholesale with the new public API, adjusting use-sites as necessary.